src/libcore/pin.rs

```diff
@@ -1,160 +1,328 @@
+//!             _pin: Pinned,
+//!         };
+//!         let mut boxed = Box::pinned(res);
+//!
+//!         let slice = NonNull::from(&boxed.data);
+//!         // we know this is safe because modifying a field doesn't move the whole struct
+//!         unsafe {
+//!             let mut_ref: Pin<&mut Self> = Pin::as_mut(&mut boxed);
+//!             Pin::get_mut_unchecked(mut_ref).slice = slice;
+//!         }
+//!         boxed
+//!     }
+//! }
+//!
+//! let unmoved = Unmovable::new("hello".to_string());
+//! // The pointer should point to the correct location,
+//! // so long as the struct hasn't moved.
+//! // Meanwhile, we are free to move the pointer around.
+//! # #[allow(unused_mut)]
+//! let mut still_unmoved = unmoved;
+//! assert_eq!(still_unmoved.slice, NonNull::from(&still_unmoved.data));
+//!
+//! // Since our type doesn't implement Unpin, this will fail to compile:
+//! // let new_unmoved = Unmovable::new("world".to_string());
+//! // std::mem::swap(&mut *still_unmoved, &mut *new_unmoved);
+//! ```
 
 #![unstable(feature = "pin", issue = "49150")]
 
 use fmt;
-use future::{Future, UnsafeFutureObj};
-use marker::{Sized, Unpin, Unsize};
-use task::{Context, Poll};
+use marker::Sized;
 use ops::{Deref, DerefMut, CoerceUnsized};
 
-/// A pinned reference.
+#[doc(inline)]
+pub use marker::Unpin;
+
+/// A pinned pointer.
 ///
-/// This type is similar to a mutable reference, except that it pins its value,
-/// which prevents it from moving out of the reference, unless it implements [`Unpin`].
+/// This is a wrapper around a kind of pointer which makes that pointer "pin" its
+/// value in place, preventing the value referenced by that pointer from being moved
+/// unless it implements [`Unpin`].
 ///
 /// See the [`pin` module] documentation for furthur explanation on pinning.
 ///
 /// [`Unpin`]: ../../std/marker/trait.Unpin.html
 /// [`pin` module]: ../../std/pin/index.html
+//
+// Note: the derives below are allowed because they all only use `&P`, so they
+// cannot move the value behind `pointer`.
 #[unstable(feature = "pin", issue = "49150")]
 #[fundamental]
-pub struct PinMut<'a, T: ?Sized + 'a> {
-    inner: &'a mut T,
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Pin<P> {
+    pointer: P,
 }
 
-#[unstable(feature = "pin", issue = "49150")]
-impl<'a, T: ?Sized + Unpin> PinMut<'a, T> {
-    /// Construct a new `PinMut` around a reference to some data of a type that
+impl<P: Deref> Pin<P>
+where
+    P::Target: Unpin,
+{
+    /// Construct a new `Pin` around a pointer to some data of a type that
     /// implements `Unpin`.
     #[unstable(feature = "pin", issue = "49150")]
-    pub fn new(reference: &'a mut T) -> PinMut<'a, T> {
-        PinMut { inner: reference }
+    #[inline(always)]
+    pub fn new(pointer: P) -> Pin<P> {
+        // Safety: the value pointed to is `Unpin`, and so has no requirements
+        // around pinning.
+        unsafe { Pin::new_unchecked(pointer) }
```

Comment:
Is this always okay? We have `impl Deref` for non-pointer types as well, and we have no safety contract at all for what you put into `P::Target`.

Reply:
If `P::Target` is `Unpin`, I'd expect basically what my comment says: that `Pin<ThingThatTargetsT>` is semantically equivalent to `ThingThatTargetsT` (no pinning invariant exists).

Reply:
I find that a rather surprising interaction. I might implement `Deref` like e.g. `TyCtxt` does, to "inherit" a whole bunch of methods from one of the members, and I will then usually not expect any interaction of that with pinning.

Or I might have a `PinnedBox` that *does* propagate pinning (that's safe, as long as `Unpin` is only conditionally implemented). Naturally I want `Deref` for `PinnedBox`.

When this new `Pin` API was introduced, it was clear that the "key point" here are the constructors that create a `Pin`. We discussed either making this manually for all types, or having a trait that lets you do this. I do not remember seeing any blanket implementation like this that relies on an *existing safe trait*, and I am pretty worried about soundness of this. I'd feel much better if we could remove this part. What is the motivation?

Reply:
IMO this new API was always just a different way to spell `PinBox`, `PinMut`, `PinRef` etc. that allowed for reusing some methods. This is one of the places where we get reuse: rather than providing safe `new_rc`, `new_arc`, `new_ref`, and `new_mut` functions where `T: Unpin`, we can have a single `new` method that supplies that bound.